Test fails with CoverageException when mixing `no_cover` with `--cov-contexts`

# Summary

When running with `--cov-context=test` tests marked with `@pytest.mark.no_cover()` fail with `coverage.exceptions.CoverageException: Cannot switch context, coverage is not started`.

## Expected vs actual result

# Reproducer

uv run --with=pytest-cov==6.1.0 pytest --cov=test_no_cover --cov-context=test ./test_no_cover.py

```python
import pytest


def test_with_coverage():
    do_print("Hi!")


@pytest.mark.no_cover()
def test_without_coverage():
    do_print("Hello!")


def do_print(message):
    print(message)
```

Full output:
```
========================================================================= test session starts ==========================================================================
platform linux -- Python 3.13.2, pytest-8.3.5, pluggy-1.5.0
rootdir: /home/wsl
plugins: cov-6.1.0
collected 2 items

test_no_cover.py .F                                                                                                                                              [100%]

=============================================================================== FAILURES ===============================================================================
________________________________________________________________________ test_without_coverage _________________________________________________________________________

self = <pytest_cov.plugin.TestContextPlugin object at 0x7f13be219a90>, item = <Function test_without_coverage>

    def pytest_runtest_call(self, item):
>       self.switch_context(item, 'run')

.cache/uv/archive-v0/cNa-as4OHY7hEnWef4tpk/lib/python3.13/site-packages/pytest_cov/plugin.py:421:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
.cache/uv/archive-v0/cNa-as4OHY7hEnWef4tpk/lib/python3.13/site-packages/pytest_cov/plugin.py:425: in switch_context
    self.cov.switch_context(context)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <coverage.control.Coverage object at 0x7f13be2e57f0>, new_context = 'test_no_cover.py::test_without_coverage|run'

    def switch_context(self, new_context: str) -> None:
        """Switch to a new dynamic context.

        `new_context` is a string to use as the :ref:`dynamic context
        <dynamic_contexts>` label for collected data.  If a :ref:`static
        context <static_contexts>` is in use, the static and dynamic context
        labels will be joined together with a pipe character.

        Coverage collection must be started already.

        .. versionadded:: 5.0

        """
        if not self._started:                           # pragma: part started
>           raise CoverageException("Cannot switch context, coverage is not started")
E           coverage.exceptions.CoverageException: Cannot switch context, coverage is not started

.cache/uv/archive-v0/cNa-as4OHY7hEnWef4tpk/lib/python3.13/site-packages/coverage/control.py:755: CoverageException
============================================================================ tests coverage ============================================================================
___________________________________________________________ coverage: platform linux, python 3.13.2-final-0 ____________________________________________________________

Name               Stmts   Miss  Cover
--------------------------------------
test_no_cover.py       8      1    88%
--------------------------------------
TOTAL                  8      1    88%
======================================================================= short test summary info ========================================================================
FAILED test_no_cover.py::test_without_coverage - coverage.exceptions.CoverageException: Cannot switch context, coverage is not started
===================================================================== 1 failed, 1 passed in 0.13s ======================================================================
```

## Versions

Python 3.13.2

```
Package    Version
---------- -------
coverage   7.8.0
iniconfig  2.1.0
packaging  24.2
pip        24.3.1
pluggy     1.5.0
pytest     8.3.5
pytest-cov 6.1.0
```

--- a/CHANGELOG.rst
+++ b/CHANGELOG.rst
@@ -2,6 +2,11 @@
 Changelog
 =========
 
+6.1.1 (2025-04-05)
+------------------
+
+* Fixed breakage that occurs when ``--cov-context`` and the ``no_cover`` marker are used together.
+
 6.1.0 (2025-04-01)
 ------------------
 

--- a/src/pytest_cov/engine.py
+++ b/src/pytest_cov/engine.py
@@ -87,6 +87,7 @@ class CovController:
         self.failed_workers = []
         self.topdir = os.fspath(Path.cwd())
         self.is_collocated = None
+        self.started = False
 
     @contextlib.contextmanager
     def ensure_topdir(self):
@@ -97,6 +98,7 @@ class CovController:
 
     @_ensure_topdir
     def pause(self):
+        self.started = False
         self.cov.stop()
         self.unset_env()
 
@@ -104,6 +106,13 @@ class CovController:
     def resume(self):
         self.cov.start()
         self.set_env()
+        self.started = True
+
+    def start(self):
+        self.started = True
+
+    def finish(self):
+        self.started = False
 
     @_ensure_topdir
     def set_env(self):
@@ -295,9 +304,12 @@ class Central(CovController):
         self.cov.start()
         self.set_env()
 
+        super().start()
+
     @_ensure_topdir
     def finish(self):
         """Stop coverage, save data to file and set the list of coverage objects to report on."""
+        super().finish()
 
         self.unset_env()
         self.cov.stop()
@@ -440,10 +452,13 @@ class DistWorker(CovController):
         )
         self.cov.start()
         self.set_env()
+        super().start()
 
     @_ensure_topdir
     def finish(self):
         """Stop coverage and send relevant info back to the master."""
+        super().finish()
+
         self.unset_env()
         self.cov.stop()
 

--- a/src/pytest_cov/plugin.py
+++ b/src/pytest_cov/plugin.py
@@ -284,7 +284,7 @@ class CovPlugin:
             self.start(engine.Central)
 
         if self.options.cov_context == 'test':
-            session.config.pluginmanager.register(TestContextPlugin(self.cov_controller.cov), '_cov_contexts')
+            session.config.pluginmanager.register(TestContextPlugin(self.cov_controller), '_cov_contexts')
 
     @pytest.hookimpl(optionalhook=True)
     def pytest_configure_node(self, node):
@@ -408,8 +408,10 @@ class CovPlugin:
 
 
 class TestContextPlugin:
-    def __init__(self, cov):
-        self.cov = cov
+    cov_controller: 'CovController'
+
+    def __init__(self, cov_controller):
+        self.cov_controller = cov_controller
 
     def pytest_runtest_setup(self, item):
         self.switch_context(item, 'setup')
@@ -421,9 +423,10 @@ class TestContextPlugin:
         self.switch_context(item, 'run')
 
     def switch_context(self, item, when):
-        context = f'{item.nodeid}|{when}'
-        self.cov.switch_context(context)
-        os.environ['COV_CORE_CONTEXT'] = context
+        if self.cov_controller.started:
+            context = f'{item.nodeid}|{when}'
+            self.cov_controller.cov.switch_context(context)
+            os.environ['COV_CORE_CONTEXT'] = context
 
 
 @pytest.fixture

--- a/tests/test_pytest_cov.py
+++ b/tests/test_pytest_cov.py
@@ -1942,6 +1942,36 @@ def test_contexts_not_supported(testdir):
     assert result.ret != 0
 
 
+def test_contexts_no_cover(testdir):
+    script = testdir.makepyfile("""
+import pytest
+
+def foobar():
+    return 1
+
+def test_with_coverage():
+    foobar()
+
+@pytest.mark.no_cover()
+def test_without_coverage():
+    foobar()
+""")
+    result = testdir.runpytest(
+        '-v',
+        '--cov-context=test',
+        '--cov=test_contexts_no_cover',
+        script,
+    )
+    result.stdout.fnmatch_lines(
+        [
+            'test_contexts_no_cover.py       8      1    88%',
+            'TOTAL                           8      1    88%',
+        ]
+    )
+    assert result.stderr.lines == []
+    assert result.ret == 0
+
+
 def test_issue_417(testdir):
     # https://github.com/pytest-dev/pytest-cov/issues/417
     whatever = testdir.maketxtfile(whatever='')

